src/main/javascript/crypto/src/libs/AttestedObject.ts

```diff
@@ -0,0 +1,166 @@
+        console.log(hash);
+
+        // TODO sign by user wallet
+        // this.signature = SignatureUtility.sign(this.unsignedEncoding, userKeys.getPrivate());
+        if (!window.ethereum){
+            throw new Error('Please install metamask before.');
+        }
+
+        const provider = new ethers.providers.Web3Provider(window.ethereum);
+        const signer = provider.getSigner();
+        if (!signer) throw new Error("Active Wallet required");
+
+        const userAddress = await signer.getAddress();
+
+        console.log('lets sign message');
+        const metamaskEnabled = await window.ethereum.enable();
+
+        if (!metamaskEnabled){
+            throw new Error("Active Wallet required");
+        }
+
+        // console.log(this.unsignedEncoding);
+
+        // let signature = await signer.signMessage(hexStringToArray(this.unsignedEncoding));
+        let signature = await signer.signMessage(ethers.utils.arrayify(hash));
+        console.log('signature');
+        console.log(signature);
+
+        if (!signature){
+            throw new Error("Cant sign data");
+        }
+
+        const ethereumHash = await ethers.utils.keccak256("\x19Ethereum Signed Message:\n" + hash.length + hash);
+        const pk = ethers.utils.recoverPublicKey(ethereumHash, signature);
+        const recoveredAddress = ethers.utils.computeAddress(ethers.utils.arrayify(pk));
+
+        console.log('recoveredAddress');
+        console.log(recoveredAddress);
+
+        //     vec.add(new DERBitString(this.signature));
+        //     this.encoding = new DERSequence(vec).getEncoded();
+        // } catch (IOException e) {
+        //     throw new RuntimeException(e);
+        // }
+        // if (!verify()) {
+        //     throw new IllegalArgumentException("The redeem request is not valid");
+        // }
+    }
+*/
+    private makeProof(attestationSecret: bigint, objectSecret: bigint, crypto: AttestationCrypto): ProofOfExponent {
+        // TODO Bob should actually verify the attestable object is valid before trying to cash it to avoid wasting gas
+        // Need to decode twice since the standard ASN1 encodes the octet string in an octet string
+        let extensions = this.att.getUnsignedAttestation().getExtensions();//.getObjectAt(0));
+        // Index in the second DER sequence is 2 since the third object in an extension is the actual value
+        // let attCom = ASN1OctetString.getInstance(extensions.getObjectAt(2)).getOctets();
+        let attCom: Uint8Array = new Uint8Array(extensions.extension.extnValue);
+        let objCom: Uint8Array = this.attestableObject.getCommitment();
+        let pok: ProofOfExponent = crypto.computeEqualityProof(uint8tohex(attCom), uint8tohex(objCom), attestationSecret, objectSecret);
+        if (!crypto.verifyEqualityProof(attCom, objCom, pok)) {
+            throw new Error("The redeem proof did not verify");
+        }
+        return pok;
+    }
+/*
+    checkValidity(){
```

Comment:
we have verified attestation and ticket ( in the stage of decode object ), do we need to verify them again?
and I not implemented sign of UseDevconTicket because I am not sure if we still need it.

Reply:
Ah, sorry. Yes, I think you are right. Because Ticket and Attestation has already been validated and we will not verify the signature in JS then the code would be superfluous here. 
Feel free to delete it

Reply:
code removed